### PR TITLE
Kinto <> Superbridge integration

### PR DESCRIPTION
**Pull request by fedealconada**
## Context
This PR is very similar to the one on [socket-DL repo](https://github.com/KintoXYZ/socket-DL/pull/1) so you can read the PR description to better understand the context.

## Description 
This PR re-uses the [Kinto helpers](https://github.com/KintoXYZ/socket-DL/blob/master/scripts/deploy/utils/kinto/kinto.ts) included in `socket-DL` to run the scripts on this repository that deploys Controllers and Vaults (might then be good the keep these helpers on a separate repo for later integrations)

It also modifies the scripts so hooks can be deployed accordingly.

### Hooks

Following the [integration guide](https://sockettech.notion.site/SuperBridge-Integration-Docs-44b61ffcfcdb4e2db83dd89bfb985709?pvs=25#a916aeec1f93432d8d98346deb28a578), this PR implements 2 different hooks that Kinto needs:

- SenderHook.sol: this hook is deployed on all vault chains (all chains Kinto will support except for Kinto). This hook is needed as it is the one that will be passing the `msg.sender`, which is the user bridging assets from the non-Kinto chain into Kinto.
- KintoHook.sol: this hook will be deployed on all Controllers (only on Kinto) and performs the following checks:
  - on `srcPreHookCall`, which is called when a user on Kinto wants to "withdraw" (bridge out), it checks if the user is KYC'd. If not, it reverts. 
  - on `dstPreHookCall`, which is called when a user wants to bridge in assets into Kinto, it checks if the _original sender_ (the one that initiated the transaction on the vault chain) is an address included in `funderWhitelist` of the user's KintoWallet. If not, it reverts. The _original sender_ is passed as an encoded param through the `SenderHook`.

_both hooks also inherit the pre-built [LimitHook.sol](https://sockettech.notion.site/SuperBridge-Integration-Docs-44b61ffcfcdb4e2db83dd89bfb985709?pvs=25#8845b42ff8064c27bb5b7938f7f5c6b3)_

### CODE REVIEW

File: contracts/hooks/KintoHook.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    function walletTs(address) external view returns (uint256);
+}
+
+interface IKintoWallet {
+    function isFunderWhitelisted(address) external view returns (bool);
+}
+
+/**
+ * @title Kinto Hook
+ * @notice TODO
```
Comment (ylv-io) [inline contracts/hooks/KintoHook.sol]: TODO?

File: contracts/hooks/KintoHook.sol [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+contract KintoHook is LimitHook {
+    IKintoID public immutable kintoID;
+    IKintoFactory public immutable kintoFactory;
+
+    error KYCRequired();
+    error SenderNotAllowed();
+
+    /**
+     * @notice Constructor for creating a Kinto Hook
+     * @param owner_ Owner of this contract.
```
Comment (ylv-io) [inline contracts/hooks/KintoHook.sol:39]: Other params?

File: contracts/hooks/KintoHook.sol [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    function srcPreHookCall(
+        SrcPreHookCallParams memory params_
+    )
+        public
+        override
+        isVaultOrController
+        returns (TransferInfo memory transferInfo, bytes memory postHookData)
+    {
+        if (!kintoID.isKYC(params_.msgSender)) revert KYCRequired();
```
Comment (rrecuero) [inline contracts/hooks/KintoHook.sol]: The sender is mostly going to be a wallet here right?

We should check that the sender is a valid wallet and that its first signer is KYC, can we also check that the destination withdrawal address is also part of the funder whitelist?

**Review by rrecuero — CHANGES_REQUESTED** [review]
Need to change the srcPreHook

**Review by ylv-io — APPROVED** [review]
🚢

**Pull request merged**